luigi/contrib/external_program.py

```diff
@@ -0,0 +1,159 @@
+import logging
+import os
+import signal
+import subprocess
+import sys
+import tempfile
+
+import luigi
+
+logger = logging.getLogger('luigi-interface')
+
+
+class ExternalProgramTask(luigi.Task):
+    """
+    Template task for running an external program in a subprocess
+
+    The program is run using :py:class:`subprocess.Popen`, with ``args`` passed
+    as a list, generated by :py:meth:`program_args` (where the first element should
+    be the executable). See :py:class:`subprocess.Popen` for details.
+
+    Your must override :py:meth:`program_args` to specify the arguments you want,
+    and you can optionally override :py:meth:`program_environment` if you want to
+    control the environment variables (see :py:class:`ExternalPythonProgramTask`
+    for an example).
+    """
+
+    def program_args(self):
+        """
+        Subclass this method to map your task parameters to the program arguments
+
+        :return: list to pass as ``args`` to :py:class:`subprocess.Popen`
+        """
+        raise NotImplementedError
+
+    def program_environment(self):
+        """
+        Subclass this method to control environment variables for the program
+
+        :return: dict mapping environment variable names to values
+        """
+        env = os.environ.copy()
+        return env
+
+    def _clean_output_file(self, file):
+        file.seek(0)
+        return ''.join(map(lambda s: s.decode('utf-8'), file.readlines()))
+
+    def run(self):
+        args = list(map(str, self.program_args()))
+
+        logger.info('Running command: %s', ' '.join(args))
+        tmp_stdout, tmp_stderr = tempfile.TemporaryFile(), tempfile.TemporaryFile()
+        env = self.program_environment()
+        proc = subprocess.Popen(
+            args,
+            env=env,
+            stdout=tmp_stdout,
+            stderr=tmp_stderr
+        )
+
+        try:
+            with ExternalProgramRunContext(proc):
+                proc.wait()
+
+            stdout = self._clean_output_file(tmp_stdout)
+            stderr = self._clean_output_file(tmp_stderr)
+
+            if stdout:
+                logger.info('Program stdout:\n{}'.format(stdout))
```

Comment:
True. This behavior is "inherited" from [luigi.contrib.spark][1] (which in turn I think inherited it from [luigi.contrib.hadoop][2]), and I didn't consider it, but this could perhaps be improved if we move to plumbum?

[1]: https://github.com/spotify/luigi/blob/master/luigi/contrib/spark.py#L241
[2]: https://github.com/spotify/luigi/blob/master/luigi/contrib/hadoop.py#L280